src/test/validation_tests.cpp

```diff
@@ -58,4 +59,97 @@ BOOST_AUTO_TEST_CASE(special_tx_validation_test)
     RegtestDeactivateSapling();
 }
 
+void test_simple_sapling_invalidity(CMutableTransaction& tx)
+{
+    CAmount nDummyValueOut;
+    {
+        CMutableTransaction newTx(tx);
+        CValidationState state;
+
+        BOOST_CHECK(!CheckTransaction(newTx, false, false, state, false));
+        BOOST_CHECK(state.GetRejectReason() == "bad-txns-vin-empty");
+    }
+    {
+        CMutableTransaction newTx(tx);
+        CValidationState state;
+
+        newTx.sapData->vShieldedSpend.emplace_back();
+        newTx.sapData->vShieldedSpend[0].nullifier = GetRandHash();
+
+        BOOST_CHECK(!CheckTransaction(newTx, false, false, state, false));
+        BOOST_CHECK(state.GetRejectReason() == "bad-txns-vout-empty");
+    }
+    {
+        // Ensure that nullifiers are never duplicated within a transaction.
+        CMutableTransaction newTx(tx);
+        CValidationState state;
+
+        newTx.sapData->vShieldedSpend.emplace_back();
+        newTx.sapData->vShieldedSpend[0].nullifier = GetRandHash();
+
+        newTx.sapData->vShieldedOutput.emplace_back();
+
+        newTx.sapData->vShieldedSpend.emplace_back();
+        newTx.sapData->vShieldedSpend[1].nullifier = newTx.sapData->vShieldedSpend[0].nullifier;
+
+        BOOST_CHECK(!SaplingValidation::CheckTransactionWithoutProofVerification(newTx, state, nDummyValueOut));
+        BOOST_CHECK(state.GetRejectReason() == "bad-spend-description-nullifiers-duplicate");
+
+        newTx.sapData->vShieldedSpend[1].nullifier = GetRandHash();
+
+        BOOST_CHECK(SaplingValidation::CheckTransactionWithoutProofVerification(newTx, state, nDummyValueOut));
+    }
+    {
+        CMutableTransaction newTx(tx);
+        CValidationState state;
+
+        // Create a coinbase transaction
+        CTxIn vin;
+        vin.prevout = COutPoint();
+        newTx.vin.emplace_back(vin);
+        CTxOut vout;
+        vout.nValue = 2;
+        newTx.vout.emplace_back(vout);
+
+        newTx.sapData->vShieldedSpend.emplace_back();
+
+        BOOST_CHECK(!CheckTransaction(newTx, false, false, state, false, false, true));
+        BOOST_CHECK(state.GetRejectReason() == "bad-txns-invalid-sapling");
+    }
+    {
+        CMutableTransaction newTx(tx);
+        CValidationState state;
+
+        // Create a coinstake transaction
+        CTxIn vin;
+        vin.prevout = COutPoint(UINT256_ZERO, 0);
+        newTx.vin.emplace_back(vin);
+        CTxOut vout;
+        vout.nValue = 2;
+        newTx.vout.emplace_back();
```

Comment:
need to set the vout's `nValue` to `0` prior to emplace here otherwise `IsCoinStake()` returns false (default initialization sets the `nValue` to `-1`.

```diff
Index: src/test/validation_tests.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/validation_tests.cpp b/src/test/validation_tests.cpp
--- a/src/test/validation_tests.cpp	(revision 8d0e73363d7b70f77121dbc9db70bc4787cd3ddd)
+++ b/src/test/validation_tests.cpp	(date 1607462459546)
@@ -125,8 +125,9 @@
         vin.prevout = COutPoint(UINT256_ZERO, 0);
         newTx.vin.emplace_back(vin);
         CTxOut vout;
+        vout.nValue = 0;
+        newTx.vout.emplace_back(vout);
         vout.nValue = 2;
-        newTx.vout.emplace_back();
         newTx.vout.emplace_back(vout);
 
         newTx.sapData->vShieldedSpend.emplace_back();
```